// datapackage/lib/config.js
'use strict';

// Module API

var IS_BROWSER = typeof window !== 'undefined';
var TABULAR_FORMATS = ['csv', 'tsv', 'xls', 'xlsx'];
var DEFAULT_DATA_PACKAGE_PROFILE = 'data-package';
var DEFAULT_RESOURCE_PROFILE = 'data-resource';
var DEFAULT_RESOURCE_ENCODING = 'utf-8';
var DEFAULT_FIELD_TYPE = 'string';
var DEFAULT_FIELD_FORMAT = 'default';
var DEFAULT_MISSING_VALUES = [''];
var DEFAULT_DIALECT = {
  delimiter: ',',
  doubleQuote: true,
  lineTerminator: '\r\n',
  quoteChar: '"',
  skipInitialSpace: true,
  header: true,
  caseSensitiveHeader: false

  // System

};module.exports = {
  IS_BROWSER: IS_BROWSER,
  TABULAR_FORMATS: TABULAR_FORMATS,
  DEFAULT_DATA_PACKAGE_PROFILE: DEFAULT_DATA_PACKAGE_PROFILE,
  DEFAULT_RESOURCE_PROFILE: DEFAULT_RESOURCE_PROFILE,
  DEFAULT_RESOURCE_ENCODING: DEFAULT_RESOURCE_ENCODING,
  DEFAULT_FIELD_TYPE: DEFAULT_FIELD_TYPE,
  DEFAULT_FIELD_FORMAT: DEFAULT_FIELD_FORMAT,
  DEFAULT_MISSING_VALUES: DEFAULT_MISSING_VALUES,
  DEFAULT_DIALECT: DEFAULT_DIALECT
};
